lib/src/container/store.rs

```diff
@@ -0,0 +1,383 @@
+                return Err(anyhow!("containers-policy.json specifies a default of `insecureAcceptAnything`; refusing usage"));
+            }
+            SignatureSource::OstreeRemote(_) => {
+                return Err(anyhow!(
+                    "Cannot currently verify layered containers via ostree remote"
+                ));
+            }
+            _ => {}
+        }
+
+        let (manifest_digest, manifest_bytes) = self.proxy.fetch_manifest().await?;
+        let manifest: oci::Manifest = serde_json::from_slice(&manifest_bytes)?;
+        let new_imageid = manifest.imageid();
+
+        // Query for previous stored state
+        let (previous_manifest_digest, previous_imageid) =
+            if let Some(merge_commit) = self.repo.resolve_rev(&self.ostree_ref, true)? {
+                let (merge_commit_obj, _) = self.repo.load_commit(merge_commit.as_str())?;
+                let commit_meta = &merge_commit_obj.child_value(0);
+                let commit_meta = ostree::glib::VariantDict::new(Some(commit_meta));
+                let previous_digest: String =
+                    commit_meta.lookup(META_MANIFEST_DIGEST)?.ok_or_else(|| {
+                        anyhow!("Missing {} metadata on merge commit", META_MANIFEST_DIGEST)
+                    })?;
+                // If the manifest digests match, we're done.
+                if previous_digest == manifest_digest {
+                    return Ok(PrepareResult::AlreadyPresent(merge_commit.to_string()));
+                }
+                // Failing that, if they have the same imageID, we're also done.
+                let previous_manifest = manifest_from_commitmeta(&commit_meta)?;
+                if previous_manifest.imageid() == new_imageid {
+                    return Ok(PrepareResult::AlreadyPresent(merge_commit.to_string()));
+                }
+                (
+                    Some(previous_digest),
+                    Some(previous_manifest.imageid().to_string()),
+                )
+            } else {
+                (None, None)
+            };
+
+        let mut layers = manifest.layers.iter().cloned();
+        // We require a base layer.
+        let base_layer = layers.next().ok_or_else(|| anyhow!("No layers found"))?;
+        let base_layer = query_layer(&self.repo, base_layer)?;
+
+        let layers: Result<Vec<_>> = layers
+            .map(|layer| -> Result<_> { query_layer(&self.repo, layer) })
+            .collect();
+        let layers = layers?;
+
+        let imp = PreparedImport {
+            manifest,
+            manifest_digest,
+            previous_manifest_digest,
+            previous_imageid,
+            base_layer,
+            layers,
+        };
+        Ok(PrepareResult::Ready(imp))
+    }
+
+    /// Import a layered container image
+    pub async fn import(mut self, import: PreparedImport) -> Result<CompletedImport> {
+        // First download the base image (if necessary) - we need the SELinux policy
```

Comment:
This seems like a reasonable constraint, at least as a start. I do hope we don't get any usecase going against this, in the future.